File: TranspositionCipher.py


def enycrypt(str,key):
    result = ""
    temp=len(str)/len(key)
    if len(str)%len(key)!=0:
        temp+=1
    mat=[[" " for i in range(len(key))] for j in range(int(temp))]
    k=0
    for i in range(len(mat)):
        for j in range(len(mat[i])):
            if len(str)>k:
                mat[i][j]=str[k]
                k+=1
    temp=key
    temp=sorted(temp)
    for i in range(len(temp)):
        for j in range(len(key)):
            if temp[i]==key[j]:
                col=j
                break
        for j in range(len(mat)):
            result+=mat[j][col]
    return result

def decrypt(str,key):
    result = ""
    temp=len(str)/len(key)
    if len(str)%len(key)!=0:
        temp+=1
    mat=[[" " for i in range(len(key))] for j in range(int(temp))]
    temp=key
    temp=sorted(temp)
    for i in range(len(temp)):
        for j in range(len(key)):
            if temp[i]==key[j]:
                col=j
                break
        for j in range(len(mat)):
            mat[j][col]=str[i*len(mat)+j]
    for i in range(len(mat)):
        for j in range(len(mat[i])):
            result+=mat[i][j]
    return result
        
            

input_str=input("Enter the string: ")
input_key=input("Enter the key: ")
enycrypt_str=enycrypt(input_str,input_key)
print("Encrypted string: ",enycrypt_str)
decrypt_str=decrypt(enycrypt_str,input_key)
print("Decrypted string: ",decrypt_str)

File: TranspositionCipher.cpp
#include <bits/stdc++.h>
using namespace std;
void Enycrypt(string &str, string key)
{
    cout << "Given String:-  " << str << endl;
    int n = str.length();
    int m = key.length();
    int row = n / m;
    if (n % m != 0)
    {
        row++;
    }
    char mat[row][m];
    int k = 0;
    for (int i = 0; i < row; i++)
    {
        for (int j = 0; j < m; j++)
        {
            if (k < n)
            {
                mat[i][j] = str[k++];
            }
            else
            {
                mat[i][j] = ' ';
            }
        }
    }
    string temp = key;
    sort(temp.begin(), temp.end());
    str.clear();
    for (int i = 0; i < m; i++)
    {
        int index = 0;
        for (int k = 0; k < m; k++)
        {
            if (temp[i] == key[k])
            {
                key[k] = ' ';
                index = k;
                break;
            }
        }
        for (int j = 0; j < row; j++)
        {
            str.push_back(mat[j][index]);
        }
    }
    cout << "Encrypted String:-  " << str << endl;
}
void Decrypt(string &str, string key)
{
    cout << "Encrypted String:-  " << str << endl;
    int n = str.length();
    int m = key.length();
    int row = n / m;
    if (n % m != 0)
    {
        row++;
    }
    char mat[row][m];
    int k = 0;
    string temp = key;
    sort(temp.begin(), temp.end());
    for (int i = 0; i < m; i++)
    {
        int index = 0;
        for (int k = 0; k < m; k++)
        {
            if (temp[i] == key[k])
            {
                key[k] = ' ';
                index = k;
                break;
            }
        }
        for (int j = 0; j < row; j++)
        {
            mat[j][index] = str[k++];
        }
    }
    str.clear();
    for (int i = 0; i < row; i++)
    {
        for (int j = 0; j < m; j++)
        {
            str.push_back(mat[i][j]);
        }
    }
    cout << "Decrypted String:-  " << str << endl;
}
int main()
{

    string str;
    cout << "Enter the string:-  ";
    getline(cin, str);
    string key;
    cout << "Enter the key:-  ";
    cin >> key;
    bool flag = false;
    while (1)
    {
        int option;
        cout << "1. Encrypt\n2. Decrypt\n3. Exit\n";
        cin >> option;
        if (option == 1)
        {
            flag = true;
            Enycrypt(str, key);
        }
        else if (option == 2)
        {
            if (!flag)
            {
                cout << "Please Encrypt the string first\n";
                continue;
            }
            Decrypt(str, key);
            flag = false;
        }
        else if (option == 3)
        {
            break;
        }
        else
        {
            cout << "Invalid Option\n";
        }
    }
    return 0;
}